services/publisher/publisher.go

```diff
@@ -306,26 +314,62 @@ func (pub *Publisher) shouldSkipPublishing(batch *protocol.AlertBatch) (string,
 	if batch.AlertCount > 0 {
 		return "", false
 	}
-	// after this line, alert count is considered as zero but let's check metrics
-	localModeMetricsOnly := pub.cfg.Config.LocalModeConfig.Enable && len(batch.Metrics) > 0
-	if localModeMetricsOnly {
-		return "", false
-	}
-	const defaultReason = "because there are no alerts"
-	if pub.cfg.Config.LocalModeConfig.Enable {
-		return defaultReason + " or metrics and local mode is enabled", true
-	}
-	if pub.skipEmpty {
-		return defaultReason + " and skipEmpty is enabled", true
+	// after this line, alert count is considered as zero
+	const becauseThereAreNoAlerts = "because there are no alerts"
+
+	localModeConfig := &pub.cfg.Config.LocalModeConfig
+	lastReportAt := pub.lastReportAt
+	pub.botConfigMu.RLock()
+	runsBots := len(pub.botConfigs) > 0
+	pub.botConfigMu.RUnlock()
+
+	switch {
+	case localModeConfig.Enable && localModeConfig.IncludeMetrics:
+		if len(batch.Metrics) > 0 {
+			return "", false
+		}
+		return becauseThereAreNoAlerts + " or metrics in local mode", true
+
+	case localModeConfig.Enable && !localModeConfig.IncludeMetrics:
+		return becauseThereAreNoAlerts + " and metrics are skipped by default in local mode", true
+
+	case pub.skipEmpty:
+		return becauseThereAreNoAlerts + " and skipEmpty is enabled", true
+
+	case runsBots && len(batch.Metrics) > 0:
+		return "", false // do not sacrifice metrics
+
+	case runsBots && len(batch.Metrics) == 0:
+		reportDeadline := lastReportAt.Add(fastReportInterval)
+		if reportDeadline.Before(time.Now()) {
+			return "", false
+		}
+		return becauseThereAreNoAlerts + " and metrics and fast report deadline has not exceeded yet", true
+
+	case !runsBots:
+		reportDeadline := lastReportAt.Add(slowReportInterval)
```

Comment:
How do you feel about this?
```
time.Since(lastReportAt)
```

Reply:
Thanks, fixed. It looks much better now.